tools/kth_threshold.cpp

```diff
@@ -0,0 +1,211 @@
+            topk.clear();
+            float t = 0.0;
+            if (results.size() == k) {
+                t = results.back().first;
+            }
+            threshold = std::max(threshold, t);
+        }
+        for (size_t i = 0; i < terms.size(); ++i) {
+            for (size_t j = i + 1; j < terms.size(); ++j) {
+                if (pairs_set.count({terms[i], terms[j]})) {
+                    Query q;
+                    q.terms.push_back(terms[i]);
+                    q.terms.push_back(terms[j]);
+                    wand_q(make_max_scored_cursors(index, wdata, *scorer, q), index.num_docs());
+                    topk.finalize();
+                    auto results = topk.topk();
+                    topk.clear();
+                    float t = 0.0;
+                    if (results.size() == k) {
+                        t = results.back().first;
+                    }
+                    threshold = std::max(threshold, t);
+                }
+            }
+        }
+        for (size_t i = 0; i < terms.size(); ++i) {
+            for (size_t j = i + 1; j < terms.size(); ++j) {
+                for (size_t s = j + 1; s < terms.size(); ++s) {
+                    if (triples_set.count({terms[i], terms[j], terms[s]})) {
+                        Query q;
+                        q.terms.push_back(terms[i]);
+                        q.terms.push_back(terms[j]);
+                        q.terms.push_back(terms[s]);
+                        wand_q(make_max_scored_cursors(index, wdata, *scorer, q), index.num_docs());
+                        topk.finalize();
+                        auto results = topk.topk();
+                        topk.clear();
+                        float t = 0.0;
+                        if (results.size() == k) {
+                            t = results.back().first;
+                        }
+                        threshold = std::max(threshold, t);
+                    }
+                }
+            }
+        }
+        std::cout << threshold << '\n';
+    }
+}
+
+using wand_raw_index = wand_data<wand_data_raw>;
+using wand_uniform_index = wand_data<wand_data_compressed<>>;
+using wand_uniform_index_quantized = wand_data<wand_data_compressed<PayloadType::Quantized>>;
+
+int main(int argc, const char** argv)
+{
+    spdlog::drop("");
+    spdlog::set_default_logger(spdlog::stderr_color_mt(""));
+
+    std::string query_filename;
+    std::optional<std::string> pairs_filename;
+    std::optional<std::string> triples_filename;
+    size_t k;
+    std::string index_filename;
+    std::string wand_data_filename;
+    bool quantized = false;
+
+    App<arg::Index, arg::WandData<arg::WandMode::Required>, arg::Query<arg::QueryMode::Ranked>, arg::Scorer>
+        app{"A tool for performing threshold estimation using k-th score informations."};
+
+    app.add_option("-p,--pairs", pairs_filename, "Pairs filename");
+    app.add_option("-t,--triples", triples_filename, "Triples filename");
```

Comment:
First, it's not clear what these files are for, so let's document.

Second, do I understand correctly that only these passed tuples will be included in the calculation? This is very complex to use if I just want all of them. Can we have a param `-k` and use these as optional filters (and maybe name them as such, say, `--filter-pairs`? For example, I have a bunch of queries and I want thresholds including pairs and triples, then I have to prepare two additional files just to say "use all".